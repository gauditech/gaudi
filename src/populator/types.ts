type PopulatorModel = {
  model: string;
  repeat: { min: number; max: number } | number;
  data: PopulatorData[];
};

type PopulatorData = {
  field: string;
  type: "constant" | "";
};
