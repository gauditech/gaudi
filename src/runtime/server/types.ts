import { NextFunction, Request, Response } from "express";

export type ServerHttpMethods = "all" | "get" | "post" | "put" | "patch" | "delete";

export type ServerRequestHandler =
  | ((request: Request, response: Response) => void | Promise<void>)
  | ((req: Request, resp: Response, next: NextFunction) => void | Promise<void>);

export type EndpointConfig = {
  path: string;
  method: ServerHttpMethods;
  handlers: ServerRequestHandler[];
};

export type ServerMiddlewareNextFn = (err?: unknown) => void;
