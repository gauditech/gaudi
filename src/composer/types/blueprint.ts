import { Model } from "./model";

export interface Blueprint {
  models: Record<string, Model>;
}
