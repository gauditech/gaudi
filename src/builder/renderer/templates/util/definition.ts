import { Definition, PathFragment } from "@src/types/definition";

/*
 * Template helper functions for Gaudi definition
 */

export function modelDbName(modelRefKey: string, definition: Definition) {
  return (definition.models || []).find((model) => model.refKey === modelRefKey)?.dbname;
}

export function fieldDbName(modelRefKey: string, fieldRefKey: string, definition: Definition) {
  return (definition.models || [])
    .find((model) => model.refKey === modelRefKey)
    ?.fields.find((field) => field.refKey === fieldRefKey)?.dbname;
}

export function typeToDbType(type: string) {
  if (type === "serial") {
    return "Int";
  } else if (type === "integer") {
    return "Int";
  } else if (type === "text") {
    return "String";
  } else if (type === "boolean") {
    return "Boolean";
  }

  throw "UNHANDLED_GAUDI_TYPE";
}

export function buildEndpointPath(parts: PathFragment[]): string {
  return parts.reduce((accum, part) => {
    const pathPart =
      part.type === "numeric" || part.type === "text" ? `:${part.varname}` : part.value;
    return `${accum}/${pathPart}`;
  }, "");
}
