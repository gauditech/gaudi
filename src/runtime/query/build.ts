import _ from "lodash";

import { mkJoinConnection } from "./stringify";

import { getModelProp, getRef, getTargetModel } from "@src/common/refs";
import { ensureEqual } from "@src/common/utils";
import {
  DeepSelectItem,
  Definition,
  EndpointDef,
  FilterDef,
  ModelDef,
  QueryDef,
  QueryDefPath,
  SelectDef,
  SelectItem,
  SelectableItem,
  TargetDef,
} from "@src/types/definition";

// FIXME introduce Queryable with SelectableItem[]?
export type QueryTree = {
  name: string;
  alias: string;
  query: QueryDef;
  related: QueryTree[];
};

export type NamePath = string[];

/**
 * Utils
 */

export function mergePaths(paths: NamePath[]): NamePath[] {
  return _.uniqWith(paths, _.isEqual);
}

export function selectToSelectable(select: SelectDef): SelectableItem[] {
  return select.filter((s): s is SelectableItem => s.kind === "field");
}

/**
 * Endpoint query builder
 */

export type EndpointQueries = {
  parentContextQueryTrees: QueryTree[];
  targetQueryTree: QueryTree;
  responseQueryTree: QueryTree;
};

export function endpointQueries(def: Definition, endpoint: EndpointDef): EndpointQueries {
  const parentContextQueryTrees = endpoint.parentContext.map((target, index) => {
    const parentTarget = index === 0 ? null : endpoint.parentContext[index - 1];
    const namePath = parentTarget ? [parentTarget.retType, target.name] : [target.retType];
    // apply identifyWith filter
    const targetFilter = targetToFilter({ ...target, namePath });
    // apply filter from it's parent
    const filter = parentTarget
      ? applyFilterIdInContext([parentTarget.retType], targetFilter)
      : targetFilter;
    const query = queryFromParts(def, target.alias, namePath, filter, target.select);
    return buildQueryTree(def, query);
  });

  // repeat the same for target
  const e = endpoint; // just a better code formatting this way
  const parentTarget = _.last(e.parentContext);
  const namePath = parentTarget ? [parentTarget.retType, e.target.name] : [e.target.retType];

  const targetFilter =
    e.kind === "create" || e.kind === "list"
      ? undefined
      : targetToFilter({ ...e.target, namePath });

  const filter = parentTarget
    ? applyFilterIdInContext([parentTarget.retType], targetFilter)
    : targetFilter;
  const targetQuery = queryFromParts(def, e.target.alias, namePath, filter, e.target.select);
  const targetQueryTree = buildQueryTree(def, targetQuery);

  const responseQuery = queryFromParts(def, e.target.alias, namePath, filter, e.response ?? []);
  const responseQueryTree = buildQueryTree(def, responseQuery);
  return { parentContextQueryTrees, targetQueryTree, responseQueryTree };
}

function applyFilterIdInContext(namePath: NamePath, filter: FilterDef): FilterDef {
  const inFilter: FilterDef = {
    kind: "binary",
    operator: "in",
    lhs: { kind: "alias", namePath: [...namePath, "id"] },
    rhs: { kind: "variable", type: "list-integer", name: "@context_ids" },
  };
  return filter === undefined
    ? inFilter
    : {
        kind: "binary",
        operator: "and",
        lhs: filter,
        rhs: inFilter,
      };
}

function getPathRetType(def: Definition, path: NamePath): ModelDef {
  // assume it starts with model
  const { value: model } = getRef<"model">(def, path[0]);
  const ret = _.tail(path).reduce(
    (ctx, name) => getTargetModel(def.models, `${ctx.refKey}.${name}`),
    model
  );
  return ret;
}

export function selectableId(def: Definition, namePath: NamePath): SelectableItem {
  const model = getPathRetType(def, namePath);
  return {
    kind: "field",
    alias: "id",
    name: "id",
    namePath: [...namePath, "id"],
    refKey: `${model.refKey}.id`,
  };
}

/**
 * Query builder
 */

export function queryFromParts(
  def: Definition,
  name: string,
  fromPath: NamePath,
  filter: FilterDef,
  select: SelectDef
): QueryDef {
  if (select.length === 0) {
    return queryFromParts(def, name, fromPath, filter, [selectableId(def, fromPath)]);
  }
  const filterPaths = getFilterPaths(filter);
  const paths = mergePaths([fromPath, ...filterPaths]);
  const direct = getDirectChildren(paths);
  ensureEqual(direct.length, 1);
  const { value: ctx } = getRef<"model">(def, direct[0]);
  const joinPaths = processPaths(def, getRelatedPaths(paths, ctx.name), ctx, [ctx.name]);

  return {
    refKey: "N/A",
    from: { kind: "model", refKey: fromPath[0] },
    filter,
    fromPath,
    name,
    nullable: false,
    // retCardinality: "many", // FIXME,
    retType: getPathRetType(def, fromPath).refKey,
    select,
    joinPaths,
  };
}

function targetToFilter(target: TargetDef): FilterDef {
  return {
    kind: "binary",
    operator: "is",
    lhs: { kind: "alias", namePath: [...target.namePath, target.identifyWith.name] },
    rhs: {
      kind: "variable",
      type: target.identifyWith.type,
      name: target.identifyWith.paramName,
    },
  };
}

export function filterFromTargets(targets: TargetDef[]): FilterDef {
  if (!targets.length) {
    return undefined;
  }
  const [t, ...rest] = targets;
  const f = targetToFilter(t);
  if (rest.length === 0) {
    return f;
  }

  return {
    kind: "binary",
    operator: "and",
    lhs: f,
    rhs: filterFromTargets(rest),
  };
}

export function getDirectChildren(paths: NamePath[]): string[] {
  return _.chain(paths)
    .map((p) => p[0])
    .compact()
    .uniq()
    .value();
}

export function getRelatedPaths(paths: NamePath[], direct: string): NamePath[] {
  return paths.filter((path) => path[0] === direct).map(_.tail);
}

// function calculateCardinality(
//   ref: Ref<"reference" | "relation" | "query">,
//   paths: QueryDefPath[]
// ): "one" | "many" {
//   const joinsAllOnes = paths.every((p) => p.retCardinality === "one");
//   const isOne =
//     ref.kind === "reference" ||
//     (ref.kind === "relation" && ref.value.unique) ||
//     (ref.kind === "query" && ref.value.retCardinality === "one");
//   return isOne && joinsAllOnes ? "one" : "many";
// }

export function processPaths(
  def: Definition,
  paths: NamePath[],
  parentCtx: ModelDef,
  prefixNames: NamePath
): QueryDefPath[] {
  const direct = getDirectChildren(paths);
  return _.compact(
    direct.flatMap((name): QueryDefPath | null => {
      const relativeChildren = getRelatedPaths(paths, name);
      const ref = getModelProp<"field" | "reference" | "relation" | "query">(parentCtx, name);
      if (ref.kind === "field") {
        return null;
      }

      const targetCtx = getTargetModel(def.models, ref.value.refKey);
      const joinPaths = processPaths(def, relativeChildren, targetCtx, [...prefixNames, name]);
      return {
        kind: ref.kind,
        joinType: "inner",
        refKey: ref.value.refKey,
        name,
        namePath: [...prefixNames, name],
        joinPaths,
        retType: getTargetModel(def.models, ref.value.refKey).name,
        // retCardinality: calculateCardinality(ref, joinPaths),
      };
    })
  );
}

// function queryToString(def: Definition, q: QueryDef): string {}

export function getFilterPaths(filter: FilterDef): string[][] {
  switch (filter?.kind) {
    case undefined:
    case "literal":
    case "variable":
      return [];
    case "alias":
      return [[...filter.namePath]];
    case "binary": {
      return [...getFilterPaths(filter.lhs), ...getFilterPaths(filter.rhs)];
    }
  }
}

/**
 * QueryTree builder
 */

export function buildQueryTree(def: Definition, q: QueryDef): QueryTree {
  const query = { ...q, select: selectToSelectable(q.select) };
  const { value: model } = getRef<"model">(def.models, query.retType);

  return {
    name: query.name,
    alias: query.name,
    query,
    related: queriesFromSelect(def, model, q.select).map((q) => buildQueryTree(def, q)),
  };
}

function queriesFromSelect(def: Definition, model: ModelDef, select: SelectDef): QueryDef[] {
  const deep: DeepSelectItem[] = select.filter(
    (s): s is DeepSelectItem =>
      s.kind === "reference" || s.kind === "relation" || s.kind === "query"
  );
  return deep.map((s) => selectToQuery(def, model, s));
}

function selectToQuery(def: Definition, model: ModelDef, select: DeepSelectItem): QueryDef {
  const namePath = [model.name, select.name];
  return queryFromParts(
    def,
    select.alias,
    namePath,
    applyFilterIdInContext([model.name], undefined),
    [
      ...select.select.map((s) => shiftSelect(model, s, select.namePath.length - 1)),
      mkJoinConnection(model),
    ]
  );
}

function shiftSelect(model: ModelDef, select: SelectItem, by: number): SelectItem {
  const namePath = [model.name, ...select.namePath.slice(by)];
  if (select.kind === "field") {
    return {
      ...select,
      namePath,
    };
  }
  return {
    ...select,
    namePath,
    select: select.select.map((s) => shiftSelect(model, s, by)),
  };
}
