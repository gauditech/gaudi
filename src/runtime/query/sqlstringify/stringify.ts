import _ from "lodash";

import { NamePath } from "../build";

import { QueryPlan, QueryPlanExpression, QueryPlanJoin } from "./queryPlan";

import { assertUnreachable } from "@src/common/utils";

function namepathToQuotedPair(npath: NamePath): string {
  return `"${_.initial(npath).join(".")}"."${_.last(npath)}"`;
}

function namepathToQuoted(npath: NamePath): string {
  return `"${npath.join(".")}"`;
}

function exprToString(expr: QueryPlanExpression): string {
  function stringifyOp(lhs: QueryPlanExpression, rhs: QueryPlanExpression, op: string): string {
    return `${exprToString(lhs)} ${op.toUpperCase()} ${exprToString(rhs)}`;
  }
  function stringifyFn(name: string, args: QueryPlanExpression[]): string {
    return `${name}(${args.map(_.unary(exprToString)).join(", ")})`;
  }
  switch (expr.kind) {
    case "alias": {
      return namepathToQuotedPair(expr.value);
    }
    case "literal": {
      return expr.value as string; // FIXME
    }
    case "variable": {
      return `:${expr.name}`;
    }
    case "function": {
      switch (expr.fnName) {
        case "<":
        case ">":
        case ">=":
        case "<=":
        case "/":
        case "*":
        case "in":
        case "not in":
        case "and": {
          return stringifyOp(expr.args[0], expr.args[1], expr.fnName);
        }
        case "+":
        case "-":
        case "or": {
          return `(${stringifyOp(expr.args[0], expr.args[1], expr.fnName)})`;
        }
        case "is": {
          return stringifyOp(expr.args[0], expr.args[1], "=");
        }
        case "is not": {
          return stringifyOp(expr.args[0], expr.args[1], "<>");
        }
        case "length": {
          return stringifyFn("char_length", expr.args);
        }
        case "count":
        case "sum":
        case "lower":
        case "upper":
        case "now":
        case "concat": {
          return stringifyFn(expr.fnName, expr.args);
        }
        case "stringify":
        case "cryptoCompare":
        case "cryptoHash":
        case "cryptoToken":
          throw new Error(`Expression "${expr.fnName}" cannot be used in queries.`);
        default:
          return assertUnreachable(expr.fnName);
      }
    }
  }
}

function joinsToString(joins: QueryPlanJoin[]): string {
  return joins
    .map((join) => {
      switch (join.kind) {
        case "subquery": {
          const subquery = queryPlanToString(join.plan);
          return `
    JOIN (${subquery})
    AS ${namepathToQuoted(join.namePath)}
    ON ${namepathToQuotedPair(join.joinOn[0])} = ${namepathToQuotedPair(join.joinOn[1])}
    ${joinsToString(join.plan.joins)}`;
        }
        case "inline": {
          return `
    JOIN ${join.modelName.toLowerCase()} AS ${namepathToQuoted(join.namePath)}
    ON ${namepathToQuotedPair(join.joinOn[0])} = ${namepathToQuotedPair(join.joinOn[1])}
    ${joinsToString(join.joins)}
    `;
        }
      }
    })
    .join("\n");
}

export function queryPlanToString(plan: QueryPlan): string {
  const selectFrag = plan.select
    ? Object.entries(plan.select)
        .map(([alias, expr]) => `${exprToString(expr)} AS "${alias}"`)
        .join(", ")
    : "*";
  const limitFrag = plan.limit ? `LIMIT ${plan.limit}` : "";
  const orderFrag = plan.orderBy
    ? `ORDER BY ${plan.orderBy.map(
        ([path, dir]): string => `${namepathToQuotedPair(path)} ${dir}`
      )}`
    : "";
  const offsetFrag = plan.offset ? `OFFSET ${plan.offset}` : "";

  const joinFrags = joinsToString(plan.joins);
  const whereFrag = plan.filter ? `WHERE ${exprToString(plan.filter)}` : "";

  const groupByItems = plan.groupBy.map(_.unary(namepathToQuotedPair));
  const groupByFrag = groupByItems.length > 0 ? `GROUP BY ${groupByItems}` : "";

  return `
  SELECT ${selectFrag}
  FROM ${plan.entry.toLowerCase()} AS "${plan.entry}"
  ${joinFrags}
  ${whereFrag}
  ${groupByFrag}
  ${orderFrag}
  ${limitFrag}
  ${offsetFrag} 
  `;
}
