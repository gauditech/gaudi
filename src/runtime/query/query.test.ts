import _ from "lodash";

import { EndpointQueries, QueryTree, endpointQueries } from "./build";
import { queryToString } from "./stringify";

import { compile, compose, parse } from "@src/index";
import { QueryDef } from "@src/types/definition";

describe("Endpoint queries", () => {
  it("nested query", () => {
    const bp = `
    model Org {
      field slug { type text, unique }
      relation repos { from Repo, through org }
      query public_repos { from repos, filter { is_public is true }}
    }
    model Repo {
      reference org { to Org }
      field is_public { type boolean }
      field name { type text }
    }
    entrypoint Orgs {
      target model Org
      identify with slug
      entrypoint Repos {
        target relation public_repos
        response { id, name }

        get endpoint {}
      }
    }

    `;
    const def = compose(compile(parse(bp)));

    const endpoint = def.entrypoints[0].entrypoints[0].endpoints[0];
    const q = endpointQueries(def, endpoint);
    expect(q).toMatchSnapshot();
    expect(extractEndpointQueries(q)).toHaveLength(3);
    expect(extractEndpointQueries(q).map((q) => queryToString(def, q))).toMatchSnapshot();
  });
  it("chained nested query", () => {
    const bp = `
    model Org {
      field slug { type text, unique }
      relation repos { from Repo, through org }
      query public_repos { from repos, filter { is_public is true }}
      query public_issues { from public_repos.issues  }
    }
    model Repo {
      reference org { to Org }
      field is_public { type boolean }
      field name { type text }
      relation issues { from Issue, through repo }
    }
    model Issue {
      reference repo { to Repo }
      field name { type text }
    }
    entrypoint Orgs {
      target model Org
      identify with slug
      entrypoint RepoIssues {
        target relation public_issues
        response { id, name, repo { org, name, is_public } }
        get endpoint {}
      }
    }

    `;
    const def = compose(compile(parse(bp)));
    const endpoint = def.entrypoints[0].entrypoints[0].endpoints[0];
    const q = endpointQueries(def, endpoint);
    expect(q).toMatchSnapshot();
    expect(extractEndpointQueries(q)).toHaveLength(5);
    expect(extractEndpointQueries(q).map((q) => queryToString(def, q))).toMatchSnapshot();
    q;
  });
});

function extractEndpointQueries(q: EndpointQueries): QueryDef[] {
  const allTrees = [...q.parentContextQueryTrees, q.targetQueryTree, q.responseQueryTree];
  return allTrees.flatMap(extractQueryTree);
}

function extractQueryTree(qt: QueryTree): QueryDef[] {
  return [qt.query, ...qt.related.flatMap(extractQueryTree)];
}
