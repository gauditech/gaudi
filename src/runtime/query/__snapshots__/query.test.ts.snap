// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Endpoint queries chained nested query 1`] = `
{
  "parentContextQueryTrees": [
    {
      "alias": "$target_0",
      "name": "$target_0",
      "query": {
        "filter": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "slug",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "org_slug",
            "type": "text",
          },
        },
        "from": {
          "kind": "model",
          "refKey": "Org",
        },
        "fromPath": [
          "Org",
        ],
        "joinPaths": [],
        "name": "$target_0",
        "nullable": false,
        "refKey": "N/A",
        "retType": "Org",
        "select": [
          {
            "alias": "id",
            "kind": "field",
            "name": "id",
            "namePath": [
              "Org",
              "id",
            ],
            "refKey": "Org.id",
          },
        ],
      },
      "related": [],
    },
  ],
  "responseQueryTree": {
    "alias": "$target_1",
    "name": "$target_1",
    "query": {
      "filter": {
        "kind": "binary",
        "lhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "public_issues",
              "id",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "issue_id",
            "type": "integer",
          },
        },
        "operator": "and",
        "rhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "id",
            ],
          },
          "operator": "in",
          "rhs": {
            "kind": "variable",
            "name": "@context_ids",
            "type": "list-integer",
          },
        },
      },
      "from": {
        "kind": "model",
        "refKey": "Org",
      },
      "fromPath": [
        "Org",
        "public_issues",
      ],
      "joinPaths": [
        {
          "joinPaths": [],
          "joinType": "inner",
          "kind": "query",
          "name": "public_issues",
          "namePath": [
            "Org",
            "public_issues",
          ],
          "refKey": "Org.public_issues",
          "retType": "Issue",
        },
      ],
      "name": "$target_1",
      "nullable": false,
      "refKey": "N/A",
      "retType": "Issue",
      "select": [
        {
          "alias": "id",
          "kind": "field",
          "name": "id",
          "namePath": [
            "Org",
            "public_issues",
            "id",
          ],
          "refKey": "Issue.id",
        },
        {
          "alias": "name",
          "kind": "field",
          "name": "name",
          "namePath": [
            "Org",
            "public_issues",
            "name",
          ],
          "refKey": "Issue.name",
        },
      ],
    },
    "related": [
      {
        "alias": "repo",
        "name": "repo",
        "query": {
          "filter": {
            "kind": "binary",
            "lhs": {
              "kind": "alias",
              "namePath": [
                "Issue",
                "id",
              ],
            },
            "operator": "in",
            "rhs": {
              "kind": "variable",
              "name": "@context_ids",
              "type": "list-integer",
            },
          },
          "from": {
            "kind": "model",
            "refKey": "Issue",
          },
          "fromPath": [
            "Issue",
            "repo",
          ],
          "joinPaths": [
            {
              "joinPaths": [],
              "joinType": "inner",
              "kind": "reference",
              "name": "repo",
              "namePath": [
                "Issue",
                "repo",
              ],
              "refKey": "Issue.repo",
              "retType": "Repo",
            },
          ],
          "name": "repo",
          "nullable": false,
          "refKey": "N/A",
          "retType": "Repo",
          "select": [
            {
              "alias": "name",
              "kind": "field",
              "name": "name",
              "namePath": [
                "Issue",
                "repo",
                "name",
              ],
              "refKey": "Repo.name",
            },
            {
              "alias": "is_public",
              "kind": "field",
              "name": "is_public",
              "namePath": [
                "Issue",
                "repo",
                "is_public",
              ],
              "refKey": "Repo.is_public",
            },
            {
              "alias": ""__join_connection"",
              "kind": "field",
              "name": "id",
              "namePath": [
                "Issue",
                "id",
              ],
              "refKey": "Issue.id",
            },
          ],
        },
        "related": [
          {
            "alias": "org",
            "name": "org",
            "query": {
              "filter": {
                "kind": "binary",
                "lhs": {
                  "kind": "alias",
                  "namePath": [
                    "Repo",
                    "id",
                  ],
                },
                "operator": "in",
                "rhs": {
                  "kind": "variable",
                  "name": "@context_ids",
                  "type": "list-integer",
                },
              },
              "from": {
                "kind": "model",
                "refKey": "Repo",
              },
              "fromPath": [
                "Repo",
                "org",
              ],
              "joinPaths": [
                {
                  "joinPaths": [],
                  "joinType": "inner",
                  "kind": "reference",
                  "name": "org",
                  "namePath": [
                    "Repo",
                    "org",
                  ],
                  "refKey": "Repo.org",
                  "retType": "Org",
                },
              ],
              "name": "org",
              "nullable": false,
              "refKey": "N/A",
              "retType": "Org",
              "select": [
                {
                  "alias": "id",
                  "kind": "field",
                  "name": "id",
                  "namePath": [
                    "Repo",
                    "org",
                    "id",
                  ],
                  "refKey": "Org.id",
                },
                {
                  "alias": "slug",
                  "kind": "field",
                  "name": "slug",
                  "namePath": [
                    "Repo",
                    "org",
                    "slug",
                  ],
                  "refKey": "Org.slug",
                },
                {
                  "alias": ""__join_connection"",
                  "kind": "field",
                  "name": "id",
                  "namePath": [
                    "Repo",
                    "id",
                  ],
                  "refKey": "Repo.id",
                },
              ],
            },
            "related": [],
          },
        ],
      },
    ],
  },
  "targetQueryTree": {
    "alias": "$target_1",
    "name": "$target_1",
    "query": {
      "filter": {
        "kind": "binary",
        "lhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "public_issues",
              "id",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "issue_id",
            "type": "integer",
          },
        },
        "operator": "and",
        "rhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "id",
            ],
          },
          "operator": "in",
          "rhs": {
            "kind": "variable",
            "name": "@context_ids",
            "type": "list-integer",
          },
        },
      },
      "from": {
        "kind": "model",
        "refKey": "Org",
      },
      "fromPath": [
        "Org",
        "public_issues",
      ],
      "joinPaths": [
        {
          "joinPaths": [],
          "joinType": "inner",
          "kind": "query",
          "name": "public_issues",
          "namePath": [
            "Org",
            "public_issues",
          ],
          "refKey": "Org.public_issues",
          "retType": "Issue",
        },
      ],
      "name": "$target_1",
      "nullable": false,
      "refKey": "N/A",
      "retType": "Issue",
      "select": [
        {
          "alias": "id",
          "kind": "field",
          "name": "id",
          "namePath": [
            "Org",
            "public_issues",
            "id",
          ],
          "refKey": "Issue.id",
        },
      ],
    },
    "related": [],
  },
}
`;

exports[`Endpoint queries chained nested query 2`] = `
[
  "SELECT "Org".id AS id
FROM "org" AS "Org"

WHERE "Org".slug = :org_slug",
  "SELECT "Org.public_issues".id AS id
      FROM "org" AS "Org"
      JOIN
(
      SELECT "Org.public_repos.issues".id AS id, "Org.public_repos.issues".name AS name, "Org.public_repos.issues".repo_id AS repo_id, "Org".id AS "__join_connection"
            FROM "org" AS "Org"
            JOIN
        (
              SELECT "Org.repos".id AS id, "Org.repos".is_public AS is_public, "Org.repos".name AS name, "Org.repos".org_id AS org_id, "Org".id AS "__join_connection"
                    FROM "org" AS "Org"
                    JOIN
              "repo" AS "Org.repos"
              ON "Org".id = "Org.repos".org_id
                    WHERE "Org.repos".is_public = TRUE) AS "Org.public_repos"
        ON "Org".id = "Org.public_repos"."__join_connection"
        JOIN
      "issue" AS "Org.public_repos.issues"
      ON "Org.public_repos".id = "Org.public_repos.issues".repo_id
            WHERE TRUE = TRUE) AS "Org.public_issues"
ON "Org".id = "Org.public_issues"."__join_connection"
      WHERE "Org.public_issues".id = :issue_id AND "Org".id IN :@context_ids",
  "SELECT "Org.public_issues".id AS id, "Org.public_issues".name AS name
      FROM "org" AS "Org"
      JOIN
(
      SELECT "Org.public_repos.issues".id AS id, "Org.public_repos.issues".name AS name, "Org.public_repos.issues".repo_id AS repo_id, "Org".id AS "__join_connection"
            FROM "org" AS "Org"
            JOIN
        (
              SELECT "Org.repos".id AS id, "Org.repos".is_public AS is_public, "Org.repos".name AS name, "Org.repos".org_id AS org_id, "Org".id AS "__join_connection"
                    FROM "org" AS "Org"
                    JOIN
              "repo" AS "Org.repos"
              ON "Org".id = "Org.repos".org_id
                    WHERE "Org.repos".is_public = TRUE) AS "Org.public_repos"
        ON "Org".id = "Org.public_repos"."__join_connection"
        JOIN
      "issue" AS "Org.public_repos.issues"
      ON "Org.public_repos".id = "Org.public_repos.issues".repo_id
            WHERE TRUE = TRUE) AS "Org.public_issues"
ON "Org".id = "Org.public_issues"."__join_connection"
      WHERE "Org.public_issues".id = :issue_id AND "Org".id IN :@context_ids",
  "SELECT "Issue.repo".name AS name, "Issue.repo".is_public AS is_public, "Issue".id AS "__join_connection"
      FROM "issue" AS "Issue"
      JOIN
"repo" AS "Issue.repo"
ON "Issue".repo_id = "Issue.repo".id
      WHERE "Issue".id IN :@context_ids",
  "SELECT "Repo.org".id AS id, "Repo.org".slug AS slug, "Repo".id AS "__join_connection"
      FROM "repo" AS "Repo"
      JOIN
"org" AS "Repo.org"
ON "Repo".org_id = "Repo.org".id
      WHERE "Repo".id IN :@context_ids",
]
`;

exports[`Endpoint queries nested query 1`] = `
{
  "parentContextQueryTrees": [
    {
      "alias": "$target_0",
      "name": "$target_0",
      "query": {
        "filter": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "slug",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "org_slug",
            "type": "text",
          },
        },
        "from": {
          "kind": "model",
          "refKey": "Org",
        },
        "fromPath": [
          "Org",
        ],
        "joinPaths": [],
        "name": "$target_0",
        "nullable": false,
        "refKey": "N/A",
        "retType": "Org",
        "select": [
          {
            "alias": "id",
            "kind": "field",
            "name": "id",
            "namePath": [
              "Org",
              "id",
            ],
            "refKey": "Org.id",
          },
        ],
      },
      "related": [],
    },
  ],
  "responseQueryTree": {
    "alias": "$target_1",
    "name": "$target_1",
    "query": {
      "filter": {
        "kind": "binary",
        "lhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "public_repos",
              "id",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "repo_id",
            "type": "integer",
          },
        },
        "operator": "and",
        "rhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "id",
            ],
          },
          "operator": "in",
          "rhs": {
            "kind": "variable",
            "name": "@context_ids",
            "type": "list-integer",
          },
        },
      },
      "from": {
        "kind": "model",
        "refKey": "Org",
      },
      "fromPath": [
        "Org",
        "public_repos",
      ],
      "joinPaths": [
        {
          "joinPaths": [],
          "joinType": "inner",
          "kind": "query",
          "name": "public_repos",
          "namePath": [
            "Org",
            "public_repos",
          ],
          "refKey": "Org.public_repos",
          "retType": "Repo",
        },
      ],
      "name": "$target_1",
      "nullable": false,
      "refKey": "N/A",
      "retType": "Repo",
      "select": [
        {
          "alias": "id",
          "kind": "field",
          "name": "id",
          "namePath": [
            "Org",
            "public_repos",
            "id",
          ],
          "refKey": "Repo.id",
        },
        {
          "alias": "name",
          "kind": "field",
          "name": "name",
          "namePath": [
            "Org",
            "public_repos",
            "name",
          ],
          "refKey": "Repo.name",
        },
      ],
    },
    "related": [],
  },
  "targetQueryTree": {
    "alias": "$target_1",
    "name": "$target_1",
    "query": {
      "filter": {
        "kind": "binary",
        "lhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "public_repos",
              "id",
            ],
          },
          "operator": "is",
          "rhs": {
            "kind": "variable",
            "name": "repo_id",
            "type": "integer",
          },
        },
        "operator": "and",
        "rhs": {
          "kind": "binary",
          "lhs": {
            "kind": "alias",
            "namePath": [
              "Org",
              "id",
            ],
          },
          "operator": "in",
          "rhs": {
            "kind": "variable",
            "name": "@context_ids",
            "type": "list-integer",
          },
        },
      },
      "from": {
        "kind": "model",
        "refKey": "Org",
      },
      "fromPath": [
        "Org",
        "public_repos",
      ],
      "joinPaths": [
        {
          "joinPaths": [],
          "joinType": "inner",
          "kind": "query",
          "name": "public_repos",
          "namePath": [
            "Org",
            "public_repos",
          ],
          "refKey": "Org.public_repos",
          "retType": "Repo",
        },
      ],
      "name": "$target_1",
      "nullable": false,
      "refKey": "N/A",
      "retType": "Repo",
      "select": [
        {
          "alias": "id",
          "kind": "field",
          "name": "id",
          "namePath": [
            "Org",
            "public_repos",
            "id",
          ],
          "refKey": "Repo.id",
        },
      ],
    },
    "related": [],
  },
}
`;

exports[`Endpoint queries nested query 2`] = `
[
  "SELECT "Org".id AS id
FROM "org" AS "Org"

WHERE "Org".slug = :org_slug",
  "SELECT "Org.public_repos".id AS id
      FROM "org" AS "Org"
      JOIN
(
      SELECT "Org.repos".id AS id, "Org.repos".is_public AS is_public, "Org.repos".name AS name, "Org.repos".org_id AS org_id, "Org".id AS "__join_connection"
            FROM "org" AS "Org"
            JOIN
      "repo" AS "Org.repos"
      ON "Org".id = "Org.repos".org_id
            WHERE "Org.repos".is_public = TRUE) AS "Org.public_repos"
ON "Org".id = "Org.public_repos"."__join_connection"
      WHERE "Org.public_repos".id = :repo_id AND "Org".id IN :@context_ids",
  "SELECT "Org.public_repos".id AS id, "Org.public_repos".name AS name
      FROM "org" AS "Org"
      JOIN
(
      SELECT "Org.repos".id AS id, "Org.repos".is_public AS is_public, "Org.repos".name AS name, "Org.repos".org_id AS org_id, "Org".id AS "__join_connection"
            FROM "org" AS "Org"
            JOIN
      "repo" AS "Org.repos"
      ON "Org".id = "Org.repos".org_id
            WHERE "Org.repos".is_public = TRUE) AS "Org.public_repos"
ON "Org".id = "Org.public_repos"."__join_connection"
      WHERE "Org.public_repos".id = :repo_id AND "Org".id IN :@context_ids",
]
`;
